### DISCUSSION TITLE
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host when using SSL

### DISCUSSION THREAD

**Discussion by highviolet** [Kludex/uvicorn#2105] [Potential Issue]
Hi,

I'm having an issue with the above error, even when using the minmal example from the docs.
The issue is only present when I'm running with SSL using "ssl-keyfile" and "ssl-certfile", e.g.

`uvicorn main:app --ssl-certfile=cert.pem --ssl-keyfile=key.pem --port 8000 --host 0.0.0.0 --log-level trace`

I'm running uvicorn 0.23.2 with CPython 3.11.4 on Windows.

The full traceback is:
```
Exception in callback _ProactorBasePipeTransport._call_connection_lost(None)
handle: <Handle _ProactorBasePipeTransport._call_connection_lost(None)>
Traceback (most recent call last):
  File "E:\Environments\env\Lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "E:\Environments\env\Lib\asyncio\proactor_events.py", line 165, in _call_connection_lost
    self._sock.shutdown(socket.SHUT_RDWR)
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host
```

Any ideas on why this error pops up?

**Comment by fcarron** [comment]
Hi,

I have exactly the same issue.
Uvicorn 0.23.2, Python 3.11.6 , Windows 11/10

The issue happen only with SSL.  If I set backlog =1 the error happen on the second request (wait _timeout_keep_alive_=5s between the requests).

**Comment by nikolay-mansas** [comment]
Hi, I have the same mistake too. Windows 10, python 3.11.6

**Comment by 7fares** [comment]
Hi
i have the same error too, interestingly enough this error seems to happen when using Chrome, while on FireFox no such error.
although my code keeps working fine, even when this error occurs

**Reply by giannisyp** [reply]
Hi , i have the same error in my case the code keeps working fine , but the shutting down of the server freezes like forever.

**Comment by idstarz** [comment]
Got same issue on Uvicorn + SSL on Windows, on each close of connection from client side (when client closes page) server got next errors in log '[WinError 10054] An existing connection was forcibly closed by the remote host' on socket shutdown in asyncio. 
This works for me:
`asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())`
Insert this just before running uvicorn your server to change default asyncio event loop. Looks like some kind of bug of asyncio/uvicorn with default ProactorEventLoop

**Reply by jmos-cruz** [reply]
Was having the same issue and this worked for me. Thank you!

**Reply by alpharou** [reply]
> Got same issue on Uvicorn + SSL on Windows, on each close of connection from client side (when client closes page) server got next errors in log '[WinError 10054] An existing connection was forcibly closed by the remote host' on socket shutdown in asyncio. This works for me: `asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())` Insert this just before running uvicorn your server to change default asyncio event loop. Looks like some kind of bug of asyncio/uvicorn with default ProactorEventLoop

Issue still present, this fix saved me. THANK YOU!
